toolkit/scripts/precache.sh

```diff
@@ -33,58 +37,103 @@ else
     if [[ "$ID" == "mariner" ]]; then
         # Mariner version doesn't prepend the full URL, so add manually via sed on every line
         prefix="$base_url/"
     else
         prefix=""
     fi
-    repoquery $REPOQUERY_OS_ARGS --repofrompath=$repo_name,$base_url -a --qf="%{location}" | sed "s|^|$prefix|" >> $cache_working_dir/repo.txt || exit 1
+    repoquery $REPOQUERY_OS_ARGS --repofrompath=$repo_name,$base_url -a --qf="%{location}" | sed "s|^|$prefix|" >> $repo_summary_file || exit 1
 done
 
-touch "$downloaded_files"
-function download_rpm() {
+
+# For each package in the snapshot summary, find the latest version of package $1 in the repo summary file $2 and print the url
+function get_url() {
     local name="$1"
+    local repo_summary_file="$2"
     rpm_name="$name.rpm"
     # '/' is important here, otherwise we might match a similarly named package. The '/' guarantees we match the start of the name.
-    rpm_url=$(grep "/$rpm_name" $cache_working_dir/repo.txt | head -n 1)
+    rpm_url=$(grep "/$rpm_name" "$repo_summary_file" | head -n 1)
     if [[ -z "$rpm_url" ]]; then
         return
     fi
+    echo "$rpm_url"
+}
 
+touch "$downloaded_files"
+touch "$skipped_file"
+function download_rpm() {
+    local rpm_url="$1"
+    local cache_dir="$2"
+    local downloaded_files="$3"
+    local skipped_file="$4"
+    rpm_name=$(basename "$rpm_url")
     # We can't use -O with wget since that resets the timestamp. We want to use -N to only download if the file is newer so we
     # need to use cd to the cache directory and then wget the file.
-    if ! output=$(cd "$cache_dir" && wget -nv -N "$rpm_url" 2>&1); then
+    if ! output=$(cd "$cache_dir" && wget -nv -N --tries=4 --waitretry=10 "$rpm_url" 2>&1); then
         echo "Failed to pre-cache $rpm_url"
         rm -f "$cache_dir/$rpm_name"
         return
     else
         # Output will contain the url + path only if it is downloaded. If it is not downloaded, it will be an empty string. Write all files we
         # successfully downloaded to the downloaded_files file.
         if [[ -n "$output" ]]; then
-            echo "Successfully pre-cached $rpm_url"
             echo "$output" >> "$downloaded_files"
+        else
+            echo "$rpm_url" >> "$skipped_file"
         fi
     fi
 }
 
+# Background process that watches $downloaded_files and compares against the number of packages in the snapshot
+function monitor_progress() {
+    local expected_count="$1"
+    # Every 5 seconds calculate the current progress and print it
+    while true; do
+        if [[ -f "$downloaded_files" ]]; then
+            current_count=$( wc -l < "$downloaded_files")
+        fi
+        if [[ -f "$skipped_file" ]]; then
+            skipped_count=$( wc -l < "$skipped_file")
+        fi
+
+        percent=$(( (current_count + skipped_count) * 100 / expected_count ))
```

Comment:
Unfortunately if the script is killed via `ctrl+c` or similar the background process will remain.
Adding the logic to the parallel workers isn't really cleaner, since one of the requests was to limit output and I'm not aware of an easy way to pass state into the exported functions running in a sub-process (ie only print out if you are %100==0 type thing)

Reply:
Fair